core/src/main/scala/kafka/consumer/BaseConsumer.scala

```diff
@@ -54,6 +57,8 @@ class NewShinyConsumer(topic: Option[String], partitionId: Option[Int], offset:
   import org.apache.kafka.clients.consumer.KafkaConsumer
 
   val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](consumerProps)
+  val offsets = new HashMap[TopicPartition, Long]()
```

Comment:
I'm thinking we still maintain a global map, but that we do not need to `new TopicPartition()` each time when returning a record, since if we iterate per partition, we know exactly which `TopicPartition` we are iterating over right now, and in fact the global iterator over the `ConsumerRecords` is also iterating over the partitions one-by-one.